code-monkey/src/agents/service.py

```diff
@@ -0,0 +1,57 @@
+# Wraps an agent in an asyncio loop and runs it.  Creates two queues for the agent to use for input and output.
+
+import asyncio
+import logging
+from typing import Dict, Any
+
+from agents.agent import Agent
+from agents.agents import agents_by_name
+from agents.typed_queue import TypedQueue
+
+type JSONDict = Dict[str, Any]
+
+
+class AgentService:
+    def __init__(self, agent: Agent):
+        self.agent = agent
+        self.request_queue = TypedQueue[JSONDict]()
+        self.response_queue = TypedQueue[JSONDict]()
+
+    async def send_to(self, data: JSONDict) -> None:
+        await self.request_queue.put(data)
+
+    async def receive_from(self) -> JSONDict:
+        return await self.response_queue.get()
+
+    # private versions of get/put from the queues on the agent-side of things
+    async def _send(self, data: JSONDict) -> None:
+        await self.response_queue.put(data)
+
+    async def _receive(self) -> JSONDict:
+        return await self.request_queue.get()
+
+    async def send_receive(self):
+        while True:
+            data = await self._receive()
+            # assume that we're always getting a { "prompt": "..." } object, and that we're always returning a { "response": "..." } object
+            logging.info(f"Agent {self.agent.name} received prompt: {data['prompt']}")
+            response = await self.agent.run_prompt(data["prompt"])
+
+            logging.info(f"Agent {self.agent.name} responding with: {response}")
+            await self._send({"response": response})
+
+    def run(self):
+        return asyncio.create_task(self.send_receive())
+
+
+class AgentServiceDict(dict):
```

Comment:
a dict subclass that essentially gives us a factory method to use for filling in a default.  I think we only need 1 instance of each agent, since we can differentiate conversations by using different `thread_id`s.